disk update fails

**Describe the bug**
Following the new release which includes https://github.com/bpg/terraform-provider-proxmox/issues/1829 I am running into an issue when cloning a VM that has 2 disks (boot and efi) where we move the boot disk to another datastore.

In the TF_LOG I see it moves the boot disk to the correct datastore, but then continues to use its _old_ id when setting the final bits.

```
tf_http_req_body="virtio0=file%3Dshared-ssd%3Avm-108-disk-1%2Csize%3D8589934592%2Caio%3Dio_uring%2Cbackup%3D0%2Ciothread%3D1%2Cdiscard%3Dignore%2Ccache%3Dnone%2Creplicate%3D1"
```


**To Reproduce**
Steps to reproduce the behavior:

Have a VM Template with 2 disks, boot and efi, both on the same datastore.

Clone a VM from it, while setting the boot disk to be on another datastore.

See the error:

```
2025-03-25T13:43:38.406+0100 [ERROR] provider.terraform-provider-proxmox_v0.73.2: Response contains error diagnostic: @caller=/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov6/internal/diag/diagnostics.go:58 tf_req_id=672dc621-5785-b340-2ed2-68faabea6640 tf_rpc=ApplyResourceChange tf_provider_addr=registry.terraform.io/bpg/proxmox @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="disk update fails: error updating VM: received an HTTP 500 response - Reason: Internal Server Error" tf_proto_version=6.8 tf_resource_type=proxmox_virtual_environment_vm timestamp="2025-03-25T13:43:38.406+0100"
```

**Expected behavior**
It should use the correct id to set the disk attributes.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

- Single or clustered Proxmox:
- Proxmox version:
- Provider version (ideally it should be the latest version):
- Terraform/OpenTofu version:
- OS (where you run Terraform/OpenTofu from):
- Debug logs (`TF_LOG=DEBUG terraform apply`):

--- a/.github/workflows/golangci-lint.yml
+++ b/.github/workflows/golangci-lint.yml
@@ -43,4 +43,5 @@ jobs:
         uses: golangci/golangci-lint-action@v7
         with: 
           version: v2.0.2 # renovate: depName=golangci/golangci-lint datasource=github-releases
+          skip-cache: true
           args: -v --timeout=10m

--- a/Makefile
+++ b/Makefile
@@ -112,7 +112,8 @@ testacc:
 .PHONY: lint
 lint:
 	# NOTE: This target is for local runs only. For linting in CI see .github/workflows/golangci-lint.yml
-	@docker run -t --rm -v $$(pwd):/app -v ~/.cache/golangci-lint/$(GOLANGCI_LINT_VERSION):/root/.cache -w /app golangci/golangci-lint:$(GOLANGCI_LINT_VERSION) golangci-lint format run --fix
+	@docker run -t --rm -v $$(pwd):/app -v ~/.cache/golangci-lint/$(GOLANGCI_LINT_VERSION):/root/.cache -w /app golangci/golangci-lint:$(GOLANGCI_LINT_VERSION) golangci-lint fmt
+	@docker run -t --rm -v $$(pwd):/app -v ~/.cache/golangci-lint/$(GOLANGCI_LINT_VERSION):/root/.cache -w /app golangci/golangci-lint:$(GOLANGCI_LINT_VERSION) golangci-lint run
 
 .PHONY: release-build
 release-build:

--- a/fwprovider/test/resource_vm_disks_test.go
+++ b/fwprovider/test/resource_vm_disks_test.go
@@ -422,7 +422,7 @@ func TestAccResourceVMDisks(t *testing.T) {
 		
 					disk {
 						interface    = "scsi0"
-						//size = 10
+						size = 10
 					}
 				}`),
 				Check: ResourceAttributes("proxmox_virtual_environment_vm.test_disk", map[string]string{
@@ -432,7 +432,7 @@ func TestAccResourceVMDisks(t *testing.T) {
 					"disk.0.interface":         "scsi0",
 					"disk.0.iothread":          "true",
 					"disk.0.path_in_datastore": `base-\d+-disk-\d+`,
-					"disk.0.size":              "8",
+					"disk.0.size":              "10",
 					"disk.0.ssd":               "true",
 				}),
 			},
@@ -535,6 +535,85 @@ func TestAccResourceVMDisks(t *testing.T) {
 				RefreshState: true,
 			},
 		}},
+		{"clone with updating disk attributes", []resource.TestStep{{
+			Config: te.RenderConfig(`
+				resource "proxmox_virtual_environment_vm" "template" {
+					node_name = "{{.NodeName}}"
+					started   = false
+					disk {
+						datastore_id = "local-lvm"
+						interface    = "virtio0"
+						file_format  = "raw"
+						size         = 20
+					}
+				}
+				resource "proxmox_virtual_environment_vm" "clone" {
+					node_name = "{{.NodeName}}"
+					started   = false
+					clone {
+						vm_id = proxmox_virtual_environment_vm.template.vm_id
+					}
+					disk {
+						datastore_id = "local-lvm"
+						interface    = "virtio0"
+						iothread     = true
+						discard      = "on"
+						size         = 30
+						speed {
+						  iops_read = 100
+						  iops_read_burstable = 1000
+						  iops_write = 400
+						  iops_write_burstable = 800
+						}
+					}
+				}`),
+			Check: resource.ComposeTestCheckFunc(
+				ResourceAttributes("proxmox_virtual_environment_vm.clone", map[string]string{
+					"disk.0.iothread":                     "true",
+					"disk.0.discard":                      "on",
+					"disk.0.size":                         "30",
+					"disk.0.speed.0.iops_read":            "100",
+					"disk.0.speed.0.iops_read_burstable":  "1000",
+					"disk.0.speed.0.iops_write":           "400",
+					"disk.0.speed.0.iops_write_burstable": "800",
+				}),
+			),
+		}}},
+		{"clone with moving disk", []resource.TestStep{{
+			Config: te.RenderConfig(`
+				resource "proxmox_virtual_environment_vm" "template" {
+					node_name = "{{.NodeName}}"
+					started   = false
+					efi_disk {
+						datastore_id = "local-lvm"
+						type = "4m"
+					}
+					disk {
+						datastore_id = "local-lvm"
+						interface    = "virtio0"
+						file_format  = "raw"
+						size         = 20
+					}
+				}
+				resource "proxmox_virtual_environment_vm" "clone" {
+					node_name = "{{.NodeName}}"
+					started   = false
+					clone {
+						vm_id = proxmox_virtual_environment_vm.template.vm_id
+					}
+					disk {
+						datastore_id = "tank"
+						interface    = "virtio0"
+						file_format  = "raw"
+						size         = 20
+					}
+				}`),
+			Check: resource.ComposeTestCheckFunc(
+				ResourceAttributes("proxmox_virtual_environment_vm.clone", map[string]string{
+					"disk.0.datastore_id": "tank",
+				}),
+			),
+		}}},
 	}
 
 	for _, tt := range tests {

--- a/fwprovider/test/resource_vm_test.go
+++ b/fwprovider/test/resource_vm_test.go
@@ -704,50 +704,6 @@ func TestAccResourceVMClone(t *testing.T) {
 				}`),
 			ExpectError: regexp.MustCompile(`storage 'doesnotexist' does not exist`),
 		}}},
-		{"update disk speed and resize in a clone", []resource.TestStep{{
-			Config: te.RenderConfig(`
-				resource "proxmox_virtual_environment_vm" "template" {
-					node_name = "{{.NodeName}}"
-					started   = false
-					disk {
-						datastore_id = "local-lvm"
-						interface    = "virtio0"
-						file_format  = "raw"
-						size         = 20
-					}
-				}
-				resource "proxmox_virtual_environment_vm" "clone" {
-					node_name = "{{.NodeName}}"
-					started   = false
-					clone {
-						vm_id = proxmox_virtual_environment_vm.template.vm_id
-					}
-					disk {
-						datastore_id = "local-lvm"
-						interface    = "virtio0"
-						iothread     = true
-						discard      = "on"
-						size         = 30
-						speed {
-						  iops_read = 100
-						  iops_read_burstable = 1000
-						  iops_write = 400
-						  iops_write_burstable = 800
-						}
-					}
-				}`),
-			Check: resource.ComposeTestCheckFunc(
-				ResourceAttributes("proxmox_virtual_environment_vm.clone", map[string]string{
-					"disk.0.iothread":                     "true",
-					"disk.0.discard":                      "on",
-					"disk.0.size":                         "30",
-					"disk.0.speed.0.iops_read":            "100",
-					"disk.0.speed.0.iops_read_burstable":  "1000",
-					"disk.0.speed.0.iops_write":           "400",
-					"disk.0.speed.0.iops_write_burstable": "800",
-				}),
-			),
-		}}},
 	}
 
 	for _, tt := range tests {

--- a/proxmox/nodes/vms/custom_storage_device.go
+++ b/proxmox/nodes/vms/custom_storage_device.go
@@ -25,13 +25,17 @@ var StorageInterfaces = []string{"ide", "sata", "scsi", "virtio"}
 
 // CustomStorageDevice handles QEMU SATA device parameters.
 type CustomStorageDevice struct {
+	// FileVolume is the path to the storage device in format
+	// "STORAGE_ID:SIZE_IN_GiB" or "STORAGE_ID:PATH_TO_FILE".
+	// This is a required field.
+	FileVolume string `json:"file"                  url:"file"`
+
 	AIO                     *string           `json:"aio,omitempty"         url:"aio,omitempty"`
 	Backup                  *types.CustomBool `json:"backup,omitempty"      url:"backup,omitempty,int"`
 	BurstableReadSpeedMbps  *int              `json:"mbps_rd_max,omitempty" url:"mbps_rd_max,omitempty"`
 	BurstableWriteSpeedMbps *int              `json:"mbps_wr_max,omitempty" url:"mbps_wr_max,omitempty"`
 	Cache                   *string           `json:"cache,omitempty"       url:"cache,omitempty"`
 	Discard                 *string           `json:"discard,omitempty"     url:"discard,omitempty"`
-	FileVolume              string            `json:"file"                  url:"file"`
 	Format                  *string           `json:"format,omitempty"      url:"format,omitempty"`
 	IopsRead                *int              `json:"iops_rd,omitempty"     url:"iops_rd,omitempty"`
 	IopsWrite               *int              `json:"iops_wr,omitempty"     url:"iops_wr,omitempty"`

--- a/proxmox/nodes/vms/custom_storage_device.go
+++ b/proxmox/nodes/vms/custom_storage_device.go
@@ -28,7 +28,7 @@ type CustomStorageDevice struct {
 	// FileVolume is the path to the storage device in format
 	// "STORAGE_ID:SIZE_IN_GiB" or "STORAGE_ID:PATH_TO_FILE".
 	// This is a required field.
-	FileVolume string `json:"file"                  url:"file"`
+	FileVolume string `json:"file" url:"file"`
 
 	AIO                     *string           `json:"aio,omitempty"         url:"aio,omitempty"`
 	Backup                  *types.CustomBool `json:"backup,omitempty"      url:"backup,omitempty,int"`

--- a/proxmoxtf/provider/provider.go
+++ b/proxmoxtf/provider/provider.go
@@ -80,7 +80,7 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
 		csrfPreventionToken = v.(string)
 	}
 
-	//nolint:staticcheck //using GetOkExists to check if the value is set, as it can be an empty string in tests
+	//nolint:staticcheck
 	if v, ok := d.GetOkExists(mkProviderAPIToken); ok {
 		apiToken = v.(string)
 	}
@@ -89,12 +89,12 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
 		otp = v.(string)
 	}
 
-	///nolint:staticcheck //using GetOkExists to check if the value is set, as it can be an empty string in tests
+	///nolint:staticcheck
 	if v, ok := d.GetOkExists(mkProviderUsername); ok {
 		username = v.(string)
 	}
 
-	//nolint:staticcheck //using GetOkExists to check if the value is set, as it can be an empty string in tests
+	//nolint:staticcheck
 	if v, ok := d.GetOkExists(mkProviderPassword); ok {
 		password = v.(string)
 	}

--- a/proxmoxtf/resource/vm/disk/disk.go
+++ b/proxmoxtf/resource/vm/disk/disk.go
@@ -82,6 +82,17 @@ func UpdateClone(
 			)
 		}
 
+		// update other disk parameters
+		// we have to do it before moving the disk, because the disk volume and location may change
+		if currentDisk.MergeWith(*planDisk) {
+			diskUpdateBody := &vms.UpdateRequestBody{}
+			diskUpdateBody.AddCustomStorageDevice(diskInterface, *currentDisk)
+
+			if err := vmAPI.UpdateVM(ctx, diskUpdateBody); err != nil {
+				return fmt.Errorf("disk update fails: %w", err)
+			}
+		}
+
 		moveDisk := false
 
 		if *planDisk.DatastoreID != "" {
@@ -102,6 +113,8 @@ func UpdateClone(
 			if err != nil {
 				return fmt.Errorf("disk move fails: %w", err)
 			}
+			// Note: after disk move, the actual disk volume ID will be different: both datastore id *and*
+			// path in datastore will change.
 		}
 
 		if planDisk.Size.InMegabytes() > currentDisk.Size.InMegabytes() {
@@ -113,16 +126,6 @@ func UpdateClone(
 			err := vmAPI.ResizeVMDisk(ctx, diskResizeBody)
 			if err != nil {
 				return fmt.Errorf("disk resize fails: %w", err)
-			}
-		}
-
-		// update other disk parameters
-		if currentDisk.MergeWith(*planDisk) {
-			diskUpdateBody := &vms.UpdateRequestBody{}
-			diskUpdateBody.AddCustomStorageDevice(diskInterface, *currentDisk)
-
-			if err := vmAPI.UpdateVM(ctx, diskUpdateBody); err != nil {
-				return fmt.Errorf("disk update fails: %w", err)
 			}
 		}
 	}

--- a/proxmoxtf/resource/vm/disk/disk.go
+++ b/proxmoxtf/resource/vm/disk/disk.go
@@ -109,12 +109,12 @@ func UpdateClone(
 				TargetStorage:      *planDisk.DatastoreID,
 			}
 
+			// Note: after disk move, the actual disk volume ID will be different: both datastore id *and*
+			// path in datastore will change.
 			err := vmAPI.MoveVMDisk(ctx, diskMoveBody)
 			if err != nil {
 				return fmt.Errorf("disk move fails: %w", err)
 			}
-			// Note: after disk move, the actual disk volume ID will be different: both datastore id *and*
-			// path in datastore will change.
 		}
 
 		if planDisk.Size.InMegabytes() > currentDisk.Size.InMegabytes() {

